python/pyarrow/scalar.pxi

```diff
@@ -315,6 +315,26 @@ cdef class ListValue(ArrayValue):
         return result
 
 
+cdef class UnionValue(ArrayValue):
+
+    cdef void _set_array(self, const shared_ptr[CArray]& sp_array):
+        self.sp_array = sp_array
+        self.ap = <CUnionArray*> sp_array.get()
+        self.value_types = [pyarrow_wrap_data_type(self.ap.value_type(i))
+                            for i in range(self.ap.num_fields())]
```

Comment:
This is quite expensive on a per-value basis. Could these wrapped types be accessed from the parent pyarrow UnionArray?

Reply:
Oops, good point! I'm constructing the Python objects for children types once per array in the UnionType.